lib/core/src/display/shape/text/cursor.rs

```diff
@@ -63,15 +72,165 @@ impl Cursor {
     //  (https://www.freetype.org/freetype2/docs/glyphs/glyphs-3.html#section-1).
     pub fn render_position(&self, content:&mut TextComponentContent, fonts:&mut Fonts)
     -> Point2<f64>{
+        let x = Self::x_position_of_cursor_at(&self.position,content,fonts);
+        let y = self.current_line(content).start_point().y;
+        Point2::new(x,y)
+    }
+
+    fn x_position_of_cursor_at
+    (position:&CharPosition, content:&mut TextComponentContent, fonts:&mut Fonts)
+    -> f64 {
         let font     = fonts.get_render_info(content.font);
-        let mut line = self.current_line(content);
-        if self.position.column > 0 {
-            let char_index = self.position.column - 1;
-            let x          = line.get_char_x_range(char_index,font).end;
-            let y          = line.start_point().y;
-            Point2::new(x.into(),y)
+        let mut line = content.line(position.line);
+        if position.column > 0 {
+            let char_index = position.column - 1;
+            line.get_char_x_range(char_index,font).end.into()
         } else {
-            line.start_point()
+            line.start_point().x
+        }
+    }
+}
+
+
+
+// ==================
+// === Navigation ===
+// ==================
+
+/// An enum representing cursor moving step. The steps are based of possible keystrokes (arrows,
+/// Home, End, Ctrl+Home, etc.)
+#[derive(Debug,Eq,Hash,PartialEq)]
+pub enum Step {Left,Right,Up,Down,LineBegin,LineEnd,DocBegin,DocEnd}
+
+/// A struct for cursor navigation process
+pub struct CursorNavigation<'a,'b> {
+    pub content   : &'a mut TextComponentContent,
+    pub fonts     : &'b mut Fonts,
+    pub selecting : bool
+}
+
+impl<'a,'b> CursorNavigation<'a,'b> {
+    /// Jump cursor directly to given position.
+    pub fn move_cursor_to_position(&self, cursor:&mut Cursor, position:CharPosition) {
+        cursor.position = position;
+        if !self.selecting {
+            cursor.selected_to = position;
+        }
+    }
+
+    /// Move cursor by given step.
+    pub fn move_cursor(&mut self, cursor:&mut Cursor, step:&Step) {
+        let new_position = self.new_position(cursor.position,&step);
+        self.move_cursor_to_position(cursor,new_position);
+    }
+
+    /// Get cursor position at begin of given line.
+    pub fn line_begin(&self, line_index:usize) -> CharPosition {
+        CharPosition {
+            column : 0,
+            line   : line_index,
+        }
+    }
```

Comment:
I believe that this should be a smart constructor of `CharPosition` used as `CharPosition::line_begin(...)`.